### ISSUE TITLE
Schema makes ABAO cough up blood

### ISSUE THREAD

**Issue by plroebuck** [cybertk/abao#61]
So I have a SIREN-based API that I wanted to verify; I wrote a trivial JSONSchema
to get started which should have worked for anything. The schema and example validate with <http://jsonschemalint.com/draft4/#>.

Schema:
```javascript
{
    "id": "http://jsonschema.net/entity-schema#",
    "$schema": "http://json-schema.org/draft-04/schema#",
    "title": "Entity",
    "description": "schema for a SIREN hypermedia entity",
    "type": "object",
    "properties": {
        "title":      { "$ref": "#/definitions/title"      },
        "class":      { "$ref": "#/definitions/class"      },
        "properties": { "$ref": "#/definitions/properties" },
        "entities":   { "$ref": "#/definitions/entities"   },
        "actions":    { "$ref": "#/definitions/actions"    },
        "links":      { "$ref": "#/definitions/links"      }
    },
    "definitions": {
        "title": {
            "id": "#title",
            "type": "string"
        },
        "class": {
            "id": "#class",
            "type": "array",
            "items": {
                "type": "string"
            },
            "uniqueItems": true
        },
        "properties": {
            "id": "#properties",
            "type": "object"
        },
        "entities": {
            "id": "#entities",
            "type": "array",
            "items": {
                "type": "object"
            }
        },
        "actions": {
            "id": "#actions",
            "type": "array",
            "items": {
                "type": "object"
            }
        },
        "link": {
            "id": "#link",
            "type": "object",
            "properties": {
                "rel": {
                    "id": "#rel",
                    "type": "array",
                    "items": {
                      "type": "string"
                    }
                },
                "href": {
                    "id": "#href",
                    "description": "URL",
                    "type": "string"
                },
                "type": {
                    "id": "#type",
                    "description": "mediaType",
                    "type": "string"
                }
            }
        },
        "links": {
            "id": "#links",
            "type": "array",
            "items": {
                "$ref": "#/definitions/link" 
            }
        }
    }
}
```

Here's an example:
```javascript
{
  "class": [
    "api"
  ],
  "entities": [
    {
      "href": "http://127.0.0.1:8080/api/heatmaps",
      "type": "application/vnd.siren+json",
      "rel": [
        "collection",
        "self"
      ]
    }
  ],
  "links": [
    {
      "href": "http://127.0.0.1:8080/api",
      "type": "application/vnd.siren+json",
      "rel": [
        "self",
        "top"
      ]
    },
    {
      "href": "http://127.0.0.1:8080/api/help",
      "type": "text/html",
      "rel": [
        "help"
      ]
    },
    {
      "href": "http://127.0.0.1:8080/api/profile",
      "type": "application/raml+yaml",
      "rel": [
        "profile"
      ]
    }
  ]
}
```

Result:
<pre><code>
abao api.raml http://127.0.0.1:8080 --hookfiles=hooks/*_hooks.js

/usr/local/lib/node_modules/abao/node_modules/csonschema/lib/index.js:235
          throw error;
                ^
SyntaxError: Syntax error on line 1, column 1: Unexpected token s
  at nodeTransforms.Literal (/usr/local/lib/node_modules/abao/node_modules/csonschema/node_modules/cson-parser/lib/parse.js:106:15)
  at transformNode (/usr/local/lib/node_modules/abao/node_modules/csonschema/node_modules/cson-parser/lib/parse.js:203:12)
  at nodeTransforms.Value (/usr/local/lib/node_modules/abao/node_modules/csonschema/node_modules/cson-parser/lib/parse.js:83:14)
  at transformNode (/usr/local/lib/node_modules/abao/node_modules/csonschema/node_modules/cson-parser/lib/parse.js:203:12)
  at nodeTransforms.Block (/usr/local/lib/node_modules/abao/node_modules/csonschema/node_modules/cson-parser/lib/parse.js:80:14)
  at transformNode (/usr/local/lib/node_modules/abao/node_modules/csonschema/node_modules/cson-parser/lib/parse.js:203:12)
  at Object.parse (/usr/local/lib/node_modules/abao/node_modules/csonschema/node_modules/cson-parser/lib/parse.js:209:12)
  at Object.parse (/usr/local/lib/node_modules/abao/node_modules/csonschema/lib/index.js:247:23)
  at parseSchema (/usr/local/lib/node_modules/abao/lib/add-tests.js:16:25)
  at async.each.addTests.path (/usr/local/lib/node_modules/abao/lib/add-tests.js:83:36)
  at /usr/local/lib/node_modules/abao/node_modules/async/lib/async.js:122:13
  at _each (/usr/local/lib/node_modules/abao/node_modules/async/lib/async.js:46:13)
  at Object.async.each (/usr/local/lib/node_modules/abao/node_modules/async/lib/async.js:121:9)
  at /usr/local/lib/node_modules/abao/lib/add-tests.js:59:20
  at /usr/local/lib/node_modules/abao/node_modules/async/lib/async.js:122:13
  at _each (/usr/local/lib/node_modules/abao/node_modules/async/lib/async.js:46:13)
  at Object.async.each (/usr/local/lib/node_modules/abao/node_modules/async/lib/async.js:121:9)
  at addTests (/usr/local/lib/node_modules/abao/lib/add-tests.js:41:18)
  at /usr/local/lib/node_modules/abao/lib/abao.js:44:18
  at fn (/usr/local/lib/node_modules/abao/node_modules/async/lib/async.js:638:34)
  at Object._onImmediate (/usr/local/lib/node_modules/abao/node_modules/async/lib/async.js:554:34)
  at processImmediate [as _immediateCallback] (timers.js:330:15)

</code></pre>

Any ideas?

N.B. Be gentle... it's my first GitHub issue...

**Comment by cybertk** [comment]
It seems **abao** treat your schema as csonschema instead of jsonschema

**Comment by plroebuck** [comment]
OK? How do I make it not do that, since it's jsonschema?

**Comment by plroebuck** [comment]
OK, so after it started using ```jsonschema``` to do the parsing, I found two further issues.
1) If your schema contains a root schema ```id``` field, it will need to be a resolvable URL.
Hence, the auto-generated portion I started from was worthless. So I changed it from
```
    "id": "http://jsonschema.net/entity-schema#",
```
to:
```
    "id": "http://localhost/api_schemas/siren.json#",
```
and made the schema accessible externally (via Apache) as a test.
But it still came up failing with "missing schema" error.

2) To get externally resolveable URL schema to work correctly required invoking ```chai.tv4.addSchema(schema);``` prior to attempting validation; this 
dynamically loaded the remote reference, resolving my schema's URI. The ```chai2-json-schema``` module documents this, needing assertions to be synchronous. 

To sum up (I guess), as I didn't really intend to make the schemas publicly available
to begin with (and just needed them for use with ```abao```), ***I just removed ALL
```id``` fields from my schema.***

Afterwards, I was able to verify a non-error response body for the first time!

**Comment by cybertk** [comment]
:+1:
